molvis/molvisbase/src/datastructures/molecularstructure.cpp

```diff
@@ -191,6 +213,140 @@ void MolecularStructure::verifyData() const {
     }
 }
 
+namespace detail {
+
+inline double dihedralAngle(const dvec3& p0, const dvec3& p1, const dvec3& p2, const dvec3& p3) {
+    const dvec3 b1{p1 - p0};
+    const dvec3 b2{p2 - p1};
+    const dvec3 b3{p3 - p2};
+    const dvec3 c1{glm::cross(b1, b2)};
+    const dvec3 c2{glm::cross(b2, b3)};
+    return glm::atan(glm::dot(glm::cross(c1, c2), glm::normalize(b2)), glm::dot(c1, c2));
+}
+
+}  // namespace detail
+
+void MolecularStructure::computeBackboneSegments() {
+    chainSegments_.clear();
+    chainSegmentIndices_.resize(data_.atoms.positions.size());
+
+    const size_t minCount = 4;
+
+    for (auto& chain : chainResidues_) {
+        size_t invalidCount = 0u;
+        std::vector<BackboneSegment> segments;
+
+        for (auto& resIdx : chain.second) {
+            const auto res = data_.residues[resIdx];
+            const auto& resAtoms = getResidueAtoms(res.id, res.chainId);
+            BackboneSegment seg;
+            seg.resId = res.id;
+            seg.chainId = res.chainId;
+
+            for (auto& atomIdx : resAtoms) {
+                chainSegmentIndices_[atomIdx] = segments.size();
+            }
+
+            if (resAtoms.size() < minCount) {
+                segments.push_back(seg);
+                LogWarn(
+                    fmt::format("invalid residue with too few atoms (id = {}, chainId = {}, '{}')",
+                                res.id, res.chainId, res.name));
+                ++invalidCount;
+                continue;
+            }
+
+            // locate atoms
+            for (auto& atomIdx : resAtoms) {
+                const Element elem = data_.atoms.atomicNumbers[atomIdx];
+                switch (data_.atoms.atomicNumbers[atomIdx]) {
+                    case Element::N:
+                        if (!seg.n.has_value()) seg.n = atomIdx;
+                        break;
+                    case Element::C:
+                        if (data_.atoms.fullNames[atomIdx] == "CA") {
+                            if (!seg.ca.has_value()) seg.ca = atomIdx;
+                        } else {
+                            if (!seg.c.has_value()) seg.c = atomIdx;
+                        }
+                        break;
+                    case Element::O:
+                        if (!seg.o.has_value()) seg.o = atomIdx;
+                        break;
+                    default:
+                        break;
+                }
+                if (seg.valid()) break;
+            }
+
+            if (!seg.valid()) {
+                ++invalidCount;
+            }
+            segments.push_back(seg);
+        }
+        chainSegments_.emplace(chain.first, std::move(segments));
+    }
+
+    computeDihedralAngles();
```

Comment:
Avoid having functions like these that only works with side effect unless absolutely needed. This is just a renamed initialise function.